src/realm/sync/client.cpp

```diff
@@ -1644,6 +1652,35 @@ util::Future<std::string> SessionWrapper::send_test_command(std::string body)
     return m_sess->send_test_command(std::move(body));
 }
 
+void SessionWrapper::handle_pending_client_reset_acknowledgement()
+{
+    auto pending_reset = [&] {
+        auto ft = m_db->start_frozen();
+        return _impl::client_reset::has_pending_reset(ft);
+    }();
+    REALM_ASSERT(pending_reset);
+    m_sess->logger.info("Tracking pending client reset of type \"%1\" from %2", pending_reset->type,
+                        pending_reset->time);
+    util::bind_ptr<SessionWrapper> self(this);
+    async_wait_for(true, true, [self = std::move(self), pending_reset = *pending_reset](std::error_code ec) {
```

Comment:
Do you think it's a problem if more than one `async_wait_for()` calls are waiting at a time to clear the client reset? Such as if `process_pending_flx_bootstrap()` throws an exception and is run again or if a connection drops after receiving the ident message and then `Session::receive_ident_message()` is called again when the client reconnects.
Probably it won't be a problem since the write transaction will serialize the calls to `remove_pending_client_resets()` and this function is a noop if there is nothing to do.

Reply:
I think at worst it will be a no-op. We used to always remove the pending client reset whenever we got a download that advanced the upload cursor. so this should be no worse than that? I've added some extra guards and logging so that if we end up with multiple callbacks from async_wait_for, only the one that matches the pending reset info that kicked off the waiting for actually triggers removing the pending resets.